fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingByIndex.java

```diff
@@ -0,0 +1,264 @@
+        if (LOGGER.isInfoEnabled() && shouldLogBuildProgress()) {
+            final Index index = common.getIndex();
+            LOGGER.info(KeyValueLogMessage.of("Built Range",
+                    LogMessageKeys.INDEX_NAME, index.getName(),
+                    LogMessageKeys.INDEX_VERSION, index.getLastModifiedVersion(),
+                    subspaceProvider.logKey(), subspaceProvider,
+                    LogMessageKeys.NEXT_CONTINUATION, retCont,
+                    LogMessageKeys.RECORDS_SCANNED, common.getTotalRecordsScanned().get()),
+                    LogMessageKeys.INDEXER_ID, common.getUuid());
+        }
+    }
+
+    @Nonnull
+    private CompletableFuture<Void> buildIndexFromIndex(SubspaceProvider subspaceProvider, @Nonnull Subspace subspace) {
+        AtomicReference<byte[]> nextCont = new AtomicReference<>();
+
+        return AsyncUtil.whileTrue(() -> {
+            byte [] cont = nextCont.get();
+            final List<Object> additionalLogMessageKeyValues = Arrays.asList(LogMessageKeys.CALLING_METHOD, "buildIndexFromIndex",
+                    LogMessageKeys.NEXT_CONTINUATION, cont == null ? "" : cont);
+
+            // apparently, buildAsync=buildAndCommitWithRetry
+            return buildAsync( (store, recordsScanned) -> buildRangeOnly(store, cont, recordsScanned),
+                    true,
+                    additionalLogMessageKeyValues)
+                    .handle((retCont, ex) -> {
+                        if (ex == null) {
+                            maybeLogBuildProgress(subspaceProvider, retCont);
+                            if (retCont == null) {
+                                return AsyncUtil.READY_FALSE;
+                            }
+                            nextCont.set(retCont); // continuation
+                            return throttleDelay();
+                        }
+                        final RuntimeException unwrappedEx = getRunner().getDatabase().mapAsyncToSyncException(ex);
+                        if (LOGGER.isInfoEnabled()) {
+                            LOGGER.info(KeyValueLogMessage.of("possibly non-fatal error encountered building range",
+                                    LogMessageKeys.NEXT_CONTINUATION, nextCont,
+                                    LogMessageKeys.SUBSPACE, ByteArrayUtil2.loggable(subspace.pack())), ex);
+                        }
+                        throw unwrappedEx;
+                    })
+                    .thenCompose(Function.identity());
+        }, getRunner().getExecutor());
+    }
+
+    @Nonnull
+    @SuppressWarnings("unchecked")
+    private static RecordCursorResult<FDBIndexedRecord<Message>> castCursorResult(@Nonnull RecordCursorResult<?> result) {
+        // helper function
+        return (RecordCursorResult<FDBIndexedRecord<Message>>)result;
+    }
+
+    @Nullable
+    private static FDBStoredRecord<Message> castCursorResultToStoreRecord(@Nonnull RecordCursorResult<?> result) {
```

Comment:
Do these two `castCursor` methods (and the ones in the other file) still need to take a `RecordCursorResult<?>` as their argument, or can they take something better (like an `FDBIndexedRecord<Message>` in this class or an `FDBStoredRecord<Message>` in the other class) now that the signature on `iterateRangeOnly` has better generics?

(Improving the type safety of these methods was one of the reasons I suggested that change)

Reply:
++